### PR TITLE
Reset dangling forward-links

### PR DESCRIPTION
**Pull request by ineiti**
Allows the admin to remove dangling forward-links from latest block

Helps #2118 

---

🙅‍ Friendly checklist:

- [x] 0. Code comments are added (or updated) when/where needed and explain the WHY of the code.
- [x] 1. Design choices, user documentation and any additional doc are added (or updated) in READMEs.
- [x] 2. Any new behaviour is tested and small units of code that can be are unit tested.
- [x] 3. Code comments are added on tests to explain what they do.
- [x] 4. Errors are systematically wrapped with a meaningful message using `xerrors.Errorf` and the `%v` verb.
- [x] 5. Hard limit of 80 chars is always respected.
- [x] 6. Changes are backward compatible.
- [x] 7. Indentation level does not exceed 5, although 4 is already suspicious.
- [x] 8. Functions, files, and packages are kept to a manageable size and decomposed into smaller units if needed.
- [x] 9. There are no magic values.

### CODE REVIEW

File: byzcoin/service.go [hunk 1]
```diff
@@ -930,6 +930,34 @@ func (s *Service) DebugRemove(req *DebugRemoveRequest) (*DebugResponse, error) {
...
+	}
+	sb, err := s.db().GetLatestByID(req.ByzCoinID)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't get latest block: %v", err)
+	}
+	if len(sb.ForwardLink) == 0 {
+		return nil, ErrNoDangling
+	}
+	sb.ForwardLink = []*skipchain.ForwardLink{}
+	err = s.db().RemoveBlock(sb.Hash)
```
Comment (Gilthoniel) [inline byzcoin/service.go]: Why don't you remove only the forward link ? That will trigger a catch up that is not necessary.

File: byzcoin/service.go [hunk 2]
```diff
@@ -930,6 +930,34 @@ func (s *Service) DebugRemove(req *DebugRemoveRequest) (*DebugResponse, error) {
 	return &DebugResponse{}, nil
 }
 
+// ErrNoDangling indicates that the latest skipblock was valid
+var ErrNoDangling = xerrors.New("no dangling forward-links")
+
+// DebugReset removes dangling forward-links in the latest block
+func (s *Service) DebugReset(req *DebugResetRequest) (*DebugResponse, error) {
+	msg := append(req.ByzCoinID, []byte("reset")...)
+	if err := schnorr.Verify(cothority.Suite, s.ServerIdentity().Public, msg,
```
Comment (Gilthoniel) [inline byzcoin/service.go]: What prevents anyone from reusing this signature to completely empty a chain ?

File: byzcoin/service.go [hunk 3]
```diff
@@ -930,6 +930,34 @@ func (s *Service) DebugRemove(req *DebugRemoveRequest) (*DebugResponse, error) {
...
+		req.Signature); err != nil {
+		log.Error("Signature failure:", err)
+		return nil, xerrors.Errorf("verifying signature: %v", err)
+	}
+	sb, err := s.db().GetLatestByID(req.ByzCoinID)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't get latest block: %v", err)
+	}
+	if len(sb.ForwardLink) == 0 {
+		return nil, ErrNoDangling
```
Comment (Gilthoniel) [inline byzcoin/service.go]: `cothority.WrapError(ErrNoDangling)`

File: byzcoin/bcadmin/test.sh [hunk 4]
```diff
@@ -27,33 +27,44 @@ main(){
...
-    run testLinkPermission
-    run testQR
-    run testUpdateDarcDesc
-    run testResolveiid
-    run testContractValue
-    run testContractDeferred
-    run testContractConfig
-    run testContractName
+    run testReset
+#    run testDbReplay
```
Comment (Gilthoniel) [inline byzcoin/bcadmin/test.sh]: don't forget to remove the comments

File: byzcoin/bcadmin/main.go [hunk 5]
```diff
@@ -1192,6 +1192,58 @@ func debugDump(c *cli.Context) error {
 	return nil
 }
 
+func debugReset(c *cli.Context) error {
```
Comment (Gilthoniel) [inline byzcoin/bcadmin/main.go]: some errors are not wrapped

**Comment by ineiti** [comment]
Replaced by #2120

**Pull request closed without merging**